grug/malachite/src/actors/host/host.rs

```diff
@@ -0,0 +1,560 @@
+            .values(state, None, None, grug::Order::Ascending)
+        {
+            let value = value?;
+            info!(%height, %round, hash = ?value.value, "Replaying already known proposed value");
+
+            consensus.cast(ConsensusMsg::ReceivedProposedValue(
+                value,
+                ValueOrigin::Consensus,
+            ))?;
+        }
+        Ok(())
+    }
+
+    async fn restream_value(
+        &self,
+        state: &mut State,
+        _height: ctx!(Height),
+        _round: Round,
+        valid_round: Round,
+        _address: ctx!(Address),
+        block_hash: ctx!(Value::Id),
+    ) -> ActorResult<()> {
+        if let Some(data) = UNDECIDED_BLOCK.may_load(state, block_hash)? {
+            // TODO: Should we assert height, round and proposer?
+
+            // recreate fin and init parts
+            let data = ProposalData {
+                block: data,
+                valid_round,
+            };
+
+            self.stream_data(state.stream_id(), data, false).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Equivalent of prepare_proposal
+    #[tracing::instrument("get_value", skip_all, fields(height = %height, round = %round))]
+    async fn get_value(
+        &self,
+        state: &mut State,
+        height: ctx!(Height),
+        round: Round,
+        _timeout: Duration,
+        reply_to: RpcReplyPort<LocallyProposedValue<Context>>,
+    ) -> ActorResult<()> {
+        // If we have already built a block for this height and round, return it to consensus
+        // This may happen when we are restarting after a crash and replaying the WAL.
+        if let Some(value) = UNDECIDED_PROPOSALS
+            .idx
+            .proposer
+            .may_load_value(state, (*height, round.as_i64(), self.address))?
+        {
+            info!(%height, %round, hash = ?value.value, "Returning previously built value");
+
+            reply_to.send(LocallyProposedValue::new(
+                value.height,
+                value.round,
+                value.value,
+            ))?;
+
+            return Ok(());
+        }
+
+        // Crate proposal parts
+        let block = {
+            // Take txs from mempool
+            info!("Taking txs from mempool");
+            let txs = self
+                .mempool
+                .call(|reply| MempoolMsg::Take { reply }, None)
+                .await?
+                .success_or(anyhow::anyhow!("Failed to take txs from mempool"))?;
+
+            // Call prepare_proposal
+            info!("Calling prepare_proposal");
+            let txs = state.prepare_proposal(txs);
+
+            let pre_block = PreBlock::new(
+                height,
+                self.address,
+                round,
+                Timestamp::from_nanos(
+                    std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)
+                        .unwrap()
+                        .as_nanos(),
```

Comment:
For proposer based timestamp there are a few checks required.
Check the timestamp of the previous block and if higher than current time then sleep for the difference. We want the blocks to have increasing timestamps.
The other check is at proposal receive time, added a comment there.